### ISSUE TITLE
tqdm hangs up with highly irregular iterations speed

### ISSUE THREAD

**Issue by aplavin** [tqdm/tqdm#249]
This is basically the same issue as #54, but that one is closed. The problem still persists, consider this example:

``` python
for i in tqdm.trange(1000):
    if i > 500:
        sleep(10)
```

It updates first time after the first 10-sec iteration, and the second time after more than 5 minutes (!) - I didn't wait for the second update. You may say this is a highly artificial case - but it doesn't seem so: for example, when you process items, some of which are skipped (e.g. already processed, or marked to skip, ...) and others require a lot of time. I myself routinely use tqdm with iterations in the order of minutes (with another faster inner progressbar, btw) and was really surprised to see the outer progressbar not moving for hours.

I expected the example at the beginning to show `501` almost immediately and then increase by one each ten seconds.

P.S.: yes, it works correctly with `miniters=1`, but I could find that this is required only reading #54 (there is a paragraph saying like 'recommended for downloads' in the readme, but it doesn't say that the progress bar will work extremely incorrectly otherwise). And anyway, even if there is no way to make it work without additional overhead, I think `miniters=1` must be the default option, as correctness is more important than speed. And then those who need to remove every possible overhead would specify `None` or something, but as for now this case is really confusing.

**Comment by lrq3000** [comment]
This is not how `maxinterval` works (even if I would like it very much!): `maxinterval` kicks in only after the last refresh being longer than the `maxinterval` threshold and then readjust the internal parameters so that next refresh will be every 10 seconds (hopefully).

In your particular example, yes at iteration 500, you will need to basically wait 500*10 seconds, but after that, `maxinterval` will kick in and the next refresh will be 10 seconds later (at iteration 1001, 1002, 1003, etc.).

There is no way around but by setting `miniters=1`. This could be made the default behavior but I am against as this is still debatable if such extreme cases can happen in real applications. It can of course happen that some refresh gets longer than `maxinterval`, but I expect these cases to be at most a few minutes, not a few hours like in your example.

Also, I disagree that correctness in all scenarios is more important than performance: noone would use a progress bar that significantly slow down their computations. `tqdm` is meant to be usable in tight loops by default.

But we could add a better note in the readme, it's clear that the current note isn't descriptive enough.

**Comment by casperdcl** [comment]
Yes, I agree with @lrq3000

**Comment by lrq3000** [comment]
Maybe we can overcome this issue by using a monitoring thread tied to the class: it would basically sleep and wake up every `monitor_interval` (default to 10 second like default maxinterval) and check whether all the bars that have `mininterval > 1` have printed in the last `maxinterval` time (just check `self.last_print_t`), if not, then `mininterval = 1`.

If we spawn only one monitoring thread, the impact on CPU wouldn't be much (considering it would sleep much of the time except when it wakes up every `monitor_interval`).

The only issue is: is it possible to access `tqdm` class variables from a thread? If yes, then the plan should work.

I'm not very good with threading, so if someone would like to give it a try...

**Comment by aplavin** [comment]
@lrq3000 well, I think I have an idea which may help dealing with this issue and still have little overhead. If I understand correctly, the problem is calling `time()` or `clock()` on each iteration as they take about 100 ns (i.e. rate is 10 M/sec) and this is important for progress bars with a lot of iterations. So, I suggest that `tqdm` default to `miniters=1` for e.g. `total` < 10k iterations: this way the overhead from getting the time on each iteration will be less than 1 ms for the whole duration of progressbar (or maybe set an even higher threshold than 10k; I personally would choose 100k, as even for a short one-second-long progressbar the overhead would be 1%). This should get rid of the hanging issue for the huge majority of cases, because it's really unlikely to have e.g. 10k iterations each lasting for a minute (that's a whole week!).

**Comment by lrq3000** [comment]
@aplavin Yes about the time() call cost, that's the issue. About the proposed solution, unluckily I think it would be unreliable, here are two counter scenarios:

1. nested loops can have an inner loop of only a few hundreds iterations, but the outer loops will multiply that number beyond the 100K threshold easily. Of course, `tqdm` can just access all instances and compute the total of all total for all bars, but the problem is that `tqdm` has no way to know which bars are nested and which aren't (eg, parallel bars downloading stuff over internet, in which case we may reach the 100K threshold whereas we would need to keep `miniters=1`).

2. the number of iterations on its own isn't a reliable way of knowing the time it will take. The 100K figure suppose that we are measuring the number of iterations, but `tqdm` can also for example measure the size of files we processed, and in this case the 100K threshold is easily reached but not indicative whether we may trespass `maxinterval` on next iterations or not.

So we can see that we can have bars with less than 100K iterations that still need `dynamic_miniters` (tight nested loops) and on the other hand scenarios where `total > 100K` but `miniters` should be set to 1.

**Comment by lrq3000** [comment]
@aplavin Anyway, there was indeed a regression that wasn't there at the time of #54 and its fix. The regression was introduced with `maxinterval`, which was designed to be a safeguard but actually did it wrong as your example shown. So thank's for your "duplicate" issue, we could track this regression down :)

This should be fixed by #251 . If you retry your example, it will take only 6 iterations (500 * mininterval / sleep = 500 * 0.1 / 10 = 5 plus 1 iteration because the display gets done after), so this fixes the issue in your particular example and in most normal instances.

For really extreme cases (like 1M fast iterations followed by ultra-slow iterations), I tried to implement a monitoring thread in #251. This should fix any issue of this kind for good, but I need feedback about the performance impact to see if we keep it.

**Comment by aplavin** [comment]
@lrq3000, yes, totally agree on №2, I forgot about it - thought only measuring about iterations before. But as for №1 - I don't think it matters that there is like 100 usec overhead even for inner progress bars, as probably the whole object creation and stuff takes comparable time. Anyway, the solution really looks unreliable :(

And if the thread solution works ok for all cases, that would be great: then parameters like `miniters` won't be needed at all, contributing to ease of use :)

**Comment by aplavin** [comment]
And btw, out of curiosity I tried calling `time` from `cython`, and it takes just 2-3 ns (for comparison, on my machine `tqdm` has overhead of about 350 ns per iteration). I read in #54 discussion that your position for tqdm was "always pure python", but maybe it's worth considering for ultimate performance without tradeoffs like testing for time only so often.

**Comment by lrq3000** [comment]
About object creation overhead vs time() in tight nested loop, yes you're correct, but still if we can manage to reduce the computational cost somewhere it's good to take (we cannot remove the object creation overhead, it's up to the user to better define the tqdm progress bars, for example by reusing the same progress bar reset to 0 -- which cannot be done automatically because we can't know if the user is going to reuse the same bar for the next instanciations of the inner loop).

About the thread, yes indeed it could replace every interval variables, but I'd prefer not for the moment. The current implementation is just to monitor when time exceeds. The bright side of the current tqdm implementation is that it's deterministic and synchronized with the parent loop, with a thread it would be asynchronous and concurrent (so we could run into race issues, and that's not fun to manage at all!). But it can indeed be used like that if you set `miniters=1, dynamic_miniters=False, maxinterval=mininterval`, in this case the monitoring thread will manage everything. But I'm not sure it would be more efficient than with miniters, but it might.

**Comment by lrq3000** [comment]
About Cython, it's just that PyPy is already more efficient than what we can achieve with Cython. I tested Cython vs PyPy on several non-numpy applications (so just basic mathematical operations), and PyPy was always the fastest, because PyPy has some clever operation optimizations that Cython doesn't possess. Also Cython optimizes only what you expressed, whereas PyPy can do optimizations you didn't even think about and on parts of the code where it would be hardly feasible to express in Cython.

To give one example, I made a universal Reed-Solomon codec, initially in pure Python. I then converted into Cython, which took me quite a lot of time. The result was a Cython version that was 2x as fast on toy data as the pure Python running under PyPy. This was on toy data. However, when I ran against real data, the PyPy version was as fast as on the toy data, whereas the Cython version ran a bit slower than PyPy (about 1.5x). So in the end, I spent weeks optimizing for nothing. [You can check it out here](https://github.com/lrq3000/reedsolomon/blob/master/creedsolo.pyx).

Why was the PyPy version faster than the Cython version? I came to the conclusion that it was because the calling script was in pure Python: PyPy optimized everything including the calling script where the Reed-Solomon codec was called in a loop, whereas Cython only optimized the library and not the calling script. I thought it wasn't much of a deal since the calling script didn't do much computations, but it was wrong, it's important to optimize the parent loops even if they just call a cythonized library. So it means that `tqdm_cython` would be faster only to people making Cython scripts, not Python scripts.

However, we could maybe provide a quickly cythonized implementation without much effort by providing a `.pxd` file and [compile Cython in pure python mode](http://cython.readthedocs.io/en/latest/src/tutorial/pure.html). It probably won't be as optimized as a real cythonized version but I don't think any dev currently is willing to spend that much effort to cythonize `tqdm` without clear rewards.

Thank's for your code @aplavin, that's interesting to know that `libc.time` is faster.

**Comment by lrq3000** [comment]
PS: I feel like my post can be read as a no-brainer about Cython. I'm not saying that. I'm open to any PR to cythonize tqdm, but personally I won't put more effort than maybe make a `.pxd` file, because I don't think that's worth it, but if anyone want to give it a try then we will be happy to receive a PR!

**Comment by aplavin** [comment]
@lrq3000 oh, I never thought of pypy as an alternative to cython, probably because I used cython for numerical computations only, and pypy isn't an option in this case as it doesn't support the whole numpy, not even talking about less popular libraries. And overall the fact that pypy can make code more efficient doesn't really change a lot because often one needs to use cpython - it supports all the libraries and is already installed basically everywhere (in linux world).

However, I totally understand that you aren't going to put effort in cythonizing tqdm, especially after underwhelming experience with the codec :) Btw, I compiled your Reed-Solomon code and almost everything is highlighted in yellow by cython, so it is no surprise it performs worse than possible.

Well, for now I guess I'll just stick with pasting something like
```python
from tqdm import tqdm_notebook as tqdm_
tqdm = lambda *a, **k: tqdm_(*a, miniters=1, **k)
```
instead of `from tqdm import tqdm_notebook as tqdm`.

**Comment by lrq3000** [comment]
@aplavin Yes for the RS codec I only cythonized the encoding parts, not the decoding nor helper function. So the encoding is totally cythonized, there should be no yellow lines there.

But yeah Cython is faster when dealing with numpy because anyway PyPy doesn't support numpy (there was numpypy but it was dropped, although there are plans to add numpy support in the standard PyPy, it seems to be still far fetched).

So a cythonized tqdm can indeed be useful for people working with cython+numpy. I think we should implement a `.pxd` file first, and if we have good results and someone wants to go further we'll see.

**Comment by lrq3000** [comment]
Should be fixed by #251 (+ #250 for the speed regression and maxinterval wrong calculation).

**Comment by cwfparsonson** [comment]
Hi, my tqdm bar keeps hanging/freezing right towards the end of my script (when it is 99% complete), which is using multiprocessing Pool.apply_async(callback=lambda _: pbar.update(1) as in https://stackoverflow.com/a/58799918/13398561). The script works fine when I am not using tqdm, and even when I write a custom pbar class to update to sanity check that apply_async() is performing the callback correctly. I've tried setting miniters to 1 and playing with mininterval, maxinterval, and smoothing, and I've tried setting refresh to False for set_description() and set_postfix(). Any ideas what might be causing this?
